clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2245,6 +2245,44 @@ public ConsumerGroupMetadata groupMetadata() {
         return coordinator.groupMetadata();
     }
 
+    /**
+     * Alert the consumer to trigger a new rebalance by rejoining the group. This is a nonblocking call that forces
+     * the consumer to trigger a new rebalance on the next {@link #poll(Duration)} call. Note that this API does not
+     * itself initiate the rebalance, so you must still call {@link #poll(Duration)}. If a rebalance is already in
+     * progress the method will return false to indicate that it was a no-op. You can use this to decide whether to
+     * just complete the current rebalance or retry in case the current rebalance does not include the latest
+     * subscription metadata; however, in most cases it is better to determine that based on the resulting assignment
+     * of the current rebalance. Only when it is not possible to validate whether the latest metadata was used based
+     * on the received assignment should you consider retrying based on the return value.
+     * <p>
+     * You should not need to call this during normal processing, as the consumer group will manage itself
+     * automatically and rebalance when necessary. However there may be situations where the application wishes to
+     * trigger a rebalance that would otherwise not occur. For example, if some condition external and invisible to
+     * the Consumer and its group membership changes in a way that has implications for the partition assignment,
+     * this method can be used to trigger a reassignment. The kind of system change that would merit calling this
+     * method is one that would also be reflected in the {@link org.apache.kafka.clients.consumer.ConsumerPartitionAssignor.Subscription Subscription}
+     * userdata that is passed to the assignor so it can make assignment decisions based on more than group membership
+     * and metadata. If your assignor does not use this userdata, or you do not use a custom
```

Comment:
`The kind of ... more than group membership and metadata` is a bit unclear what it tries to convey. Or are you trying to say "E.g. you want to trigger a rebalance with newly encoded subscription metadata, but none of the membership or subscription topics or patterns have changed so consumer group would not automatically rebalance, you can use this function"?

Reply:
The general motivation behind this was just to avoid users calling this unnecessarily or for the wrong reasons: basically users should not use this to try and manually intervene with the group membership, as that is and should be handled entirely internally.

The partition assignment only has two types of info with which to determine the assignment, one being the cluster metadata and the other being the subscription of each member. If there is a membership/cluster metadata change, the consumer will rebalance on its own. And if nothing changes, there's no reason to trigger a rebalance as the assignment should be the same. However, if there is a change that affects the userdata encoded in the subscription, the consumer will not know about this and therefore won't trigger a rebalance -- unless you tell it to. 

Does that make sense? Again, the purpose is just to avoid users calling this except for the specific reason it exists, a change affecting subscription userdata. I'll try to think of a better way to phrase this in the javadocs without adding 1000 lines to the PR -- suggestions welcome :)

Reply:
Yeah that makes sense :) I think we are on the same page, and I agree it's a bit hard to convey the message in a straight-forward phrasing. I do not have a strong suggestion here and I think your last sentence actually is good: "E.g. if there is a change that affects the userdata encoded in the subscription", consumers would not automatically rebalance and the user may want to use this.